src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java

```diff
@@ -0,0 +1,165 @@
+import ua.chernonog.onlinebookstore.service.BookService;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class BookControllerTest {
+    private static MockMvc mockMvc;
+    @Autowired
+    private ObjectMapper objectMapper;
+    @Mock
+    private BookService bookService;
+
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext context) {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(context)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @Test
+    @WithMockUser(roles = "USER")
+    @DisplayName("Test retrieving all books")
+    void getAllBooks_WithUserRole_ShouldReturnAllBooks() throws Exception {
+        List<BookResponseDto> books = List.of(new BookResponseDto());
+        when(bookService.getAll(any(Pageable.class))).thenReturn(books);
+
+        mockMvc.perform(get(BOOKS_BASE_URL))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath(JSON_PATH_FIRST_ITEM_EXISTS).exists());
+    }
+
+    @Test
+    @WithMockUser(roles = "USER")
+    @DisplayName("Test searching books by parameters")
+    void searchBooks_WithValidParameters_ShouldReturnBooks() throws Exception {
+        List<BookResponseDto> books = List.of(new BookResponseDto());
+        BookSearchParametersDto searchParameters = new BookSearchParametersDto();
+
+        when(bookService.search(any(
+                BookSearchParametersDto.class),
+                any(Pageable.class))).thenReturn(books);
+
+        String searchJson = objectMapper.writeValueAsString(searchParameters);
+
+        mockMvc.perform(get(BOOKS_SEARCH_URL)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(searchJson))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath(JSON_PATH_FIRST_ITEM_EXISTS).exists());
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = "ADMIN")
+    @DisplayName("Test valid book creation by an admin")
+    void createBook_WithValidDataAndAdminRole_ShouldCreateBook() throws Exception {
+        CreateBookRequestDto bookDto = new CreateBookRequestDto();
+        bookDto.setTitle(TITLE);
+        bookDto.setAuthor(AUTHOR);
+        bookDto.setIsbn(ISBN);
+        bookDto.setPrice(PRICE);
+        bookDto.setDescription(DESCRIPTION);
+        bookDto.setCoverImage(COVER_IMAGE_URL);
+
+        BookResponseDto responseDto = new BookResponseDto();
+        responseDto.setId(1L);
+        responseDto.setTitle(bookDto.getTitle());
+        responseDto.setAuthor(bookDto.getAuthor());
+        responseDto.setIsbn(bookDto.getIsbn());
+        responseDto.setPrice(bookDto.getPrice());
+        responseDto.setDescription(bookDto.getDescription());
+        responseDto.setCoverImage(bookDto.getCoverImage());
+
+        when(bookService.save(any(CreateBookRequestDto.class))).thenReturn(responseDto);
+
+        objectMapper = new ObjectMapper();
```

Comment:
What is the reason to create object mapper in most methods?

Reply:
There is no any reason. I forgot to remove.